src/deepsparse/transformers/engines/nl_decoder_engine.py

```diff
@@ -0,0 +1,336 @@
+            from
+        """
+        self.kv_cache = copy.deepcopy(cache)
+
+    @staticmethod
+    def overwrite_onnx_model_inputs(
+        onnx_file_path: str,
+        sequence_length: int,
+        input_ids_length: int,
+        batch_size: int = 1,
+    ) -> Tuple[str, List[int]]:
+        """
+        Enforces the appropriate input shapes for the onnx model, as well as
+        checks whether kv cache is enabled or not.
+
+        :param onnx_file_path: The path to the onnx model file that will be
+            overwritten with the new input shapes
+        :param batch_size: The batch size to use for the input
+        :param sequence_length: The sequence length to use for the input
+        :param input_ids_length: The length of input_ids
+        :return: The path to the onnx model file that has been overwritten
+            with the new input shapes, as well as the indices of the inputs
+            that should be cached
+        """
+        model = onnx.load(onnx_file_path, load_external_data=False)
+        initializer_input_names = set([node.name for node in model.graph.initializer])
+        external_inputs = [
+            inp for inp in model.graph.input if inp.name not in initializer_input_names
+        ]
+        for external_input in external_inputs:
+            # overwrite the batch size for all the inputs
+            external_input.type.tensor_type.shape.dim[0].dim_value = batch_size
+
+            if external_input.name in ["input_ids", "positions"]:
+                external_input.type.tensor_type.shape.dim[
+                    1
+                ].dim_value = input_ids_length
+            elif external_input.name == "attention_mask":
+                external_input.type.tensor_type.shape.dim[1].dim_value = sequence_length
+            elif external_input.name.startswith(_CACHE_INPUT_NAME):
+                external_input.type.tensor_type.shape.dim[2].dim_value = (
+                    sequence_length - input_ids_length
+                )
+            else:
+                raise ValueError(
+                    f"Unexpected external input name: {external_input.name}"
+                )
+
+        _LOGGER.info(
+            "Overwriting in-place the input shapes "
+            f"of the transformer model at {onnx_file_path}"
+        )
+        save_onnx(model, onnx_file_path)
+
+        output_indices_to_be_cached = [
+            1 if inp.name.startswith("present") else 0 for inp in model.graph.output
+        ]
+
+        return onnx_file_path, output_indices_to_be_cached
+
+    def generate_token(self, logits: numpy.ndarray) -> numpy.ndarray:
+        """
+        Samples a token from the logits using the sampling temperature.
+        :param logits: the logits from the model with shape (vocab_size,)
+        :return: the sampled token
+        """
+        if self.deterministic:
+            return numpy.argmax(logits)
+
+        logits /= self.sampling_temperature
```

Comment:
will sampling temperature always be > 0; if yes could we add a pre-condition to the doctstring?